src/ripple/app/tx/impl/AMMWithdraw.cpp

```diff
@@ -0,0 +1,806 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2023 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/AMMWithdraw.h>
+
+#include <ripple/app/misc/AMMHelpers.h>
+#include <ripple/app/misc/AMMUtils.h>
+#include <ripple/basics/Number.h>
+#include <ripple/ledger/Sandbox.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/AMMCore.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/STAccount.h>
+#include <ripple/protocol/TxFlags.h>
+
+namespace ripple {
+
+NotTEC
+AMMWithdraw::preflight(PreflightContext const& ctx)
+{
+    if (!ammEnabled(ctx.rules))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const flags = ctx.tx.getFlags();
+    if (flags & tfWithdrawMask)
+    {
+        JLOG(ctx.j.debug()) << "AMM Withdraw: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    auto const amount = ctx.tx[~sfAmount];
+    auto const amount2 = ctx.tx[~sfAmount2];
+    auto const ePrice = ctx.tx[~sfEPrice];
+    auto const lpTokens = ctx.tx[~sfLPTokenIn];
+    // Valid combinations are:
+    //   LPTokens
+    //   tfWithdrawAll
+    //   Amount
+    //   tfOneAssetWithdrawAll & Amount
+    //   Amount and Amount2
+    //   Amount and LPTokens
+    //   Amount and EPrice
+    if (std::bitset<32>(flags & tfWithdrawSubTx).count() != 1)
```

Comment:
How about `std::has_single_bit` or `std::popcount`?

Reply:
What is the advantage as opposed to `std::bitset`?